corehq/apps/userreports/reports/builder/forms.py

```diff
@@ -395,44 +398,31 @@ def indicators(self, columns, filters, is_multiselect_chart_report=False):
             Each object has a "property" and "aggregation" key
         :param filters: A list of filter configuration objects
         """
-        indicators = []
+
+        def get_key(i):
+            return i['column_id'], i['type']
+
+        indicators = {}
         for column in columns:
             column_option = self.report_column_options[column['property']]
-            indicators.extend(column_option.get_indicators(column['aggregation'], is_multiselect_chart_report))
+            for indicator in column_option.get_indicators(column['aggregation'], is_multiselect_chart_report):
+                indicators.setdefault(get_key(indicator), indicator)
 
-        for filter in filters:
-            property = self.data_source_properties[filter['property']]
-            indicator = property.to_report_filter_indicator(filter)
-            indicators.append(indicator)
+        for filter_ in filters:
+            property_ = self.data_source_properties[filter_['property']]
+            indicator = property_.to_report_filter_indicator(filter_)
+            indicators.setdefault(get_key(indicator), indicator)
 
-        # remove duplicates
-        # There can be duplicates because filters and columns could be based on the same property
-        indicators_without_dups = []
-        seen_indicator_ids = set()
-        for i in indicators:
-            if (i['column_id'], i['type']) not in seen_indicator_ids:
-                indicators_without_dups.append(i)
-                seen_indicator_ids.add((i['column_id'], i['type']))
-        indicators = indicators_without_dups
-
-        return indicators
+        return list(indicators.values())
 
     def all_possible_indicators(self):
-        indicators = []
+        indicators = {}
         for column_option in self.report_column_options.values():
             for agg in column_option.aggregation_options:
-                indicators.extend(column_option.get_indicators(agg))
-
-        # Remove duplicates
-        return_list = []
-        return_list_set = set()
-        for indicator in indicators:
-            as_hashable = str(indicator)
-            if as_hashable not in return_list_set:
-                return_list.append(indicator)
-                return_list_set.add(as_hashable)
+                for indicator in column_option.get_indicators(agg):
+                    indicators.setdefault(str(indicator), indicator)
 
-        return return_list
+        return list(indicators.values())[:MAX_COLUMNS]
```

Comment:
This may not always grab the same columns because dict is an unordered data type. Could probably use [`OrderedDict`](https://docs.python.org/2/library/collections.html#collections.OrderedDict).

Reply:
Good point @millerdev . I'm not sure they need to be in the same order, but I do agree it's best to reproduce the previous behaviour exactly. I'll switch to `OrderedDict`.